core/lib/rom/command_compiler.rb

```diff
@@ -0,0 +1,260 @@
+require 'dry/core/inflector'
+require 'dry/core/cache'
+
+require 'rom/initializer'
+require 'rom/commands'
+require 'rom/command_proxy'
+
+module ROM
+  # Builds commands for relations.
+  #
+  # This class is used by repositories to automatically create commands for
+  # their relations. This is used both by `Repository#command` method and
+  # `commands` repository class macros.
+  #
+  # @api private
+  class CommandCompiler
+    extend Dry::Core::Cache
+    extend Initializer
+
+    # @api private
+    def self.registry
+      @__registry__ ||= Hash.new { |h, k| h[k] = {} }
+    end
+
+    # @!attribute [r] gateways
+    #   @return [ROM::Registry]
+    param :gateways
+
+    # @!attribute [r] relations
+    #   @return [ROM::RelationRegistry]
+    param :relations
+
+    # @!attribute [r] relations
+    #   @return [ROM::Registry]
+    param :commands
+
+    # @!attribute [r] id
+    #   @return [Symbol] The command type registry identifier
+    option :id, optional: true
+
+    # @!attribute [r] adapter
+    #   @return [Symbol] The adapter identifier ie :sql or :http
+    option :adapter, optional: true
+
+    # @!attribute [r] registry
+    #   @return [Hash] local registry where commands will be stored during compilation
+    option :registry, optional: true, default: -> { self.class.registry }
+
+    # @!attribute [r] plugins
+    #   @return [Array<Symbol>] a list of optional plugins that will be enabled for commands
+    option :plugins, optional: true, default: -> { EMPTY_ARRAY }
+
+    # @!attribute [r] meta
+    #   @return [Array<Symbol>] Meta data for a command
+    option :meta, optional: true
+
+    # Return a specific command type for a given adapter and relation AST
+    #
+    # This class holds its own registry where all generated commands are being
+    # stored
+    #
+    # CommandProxy is returned for complex command graphs as they expect root
+    # relation name to be present in the input, which we don't want to have
+    # in repositories. It might be worth looking into removing this requirement
+    # from rom core Command::Graph API.
+    #
+    # @overload [](type, adapter, ast, plugins, meta)
+    #   @param type [Symbol] The type of command
+    #   @param adapter [Symbol] The adapter identifier
+    #   @param ast [Array] The AST representation of a relation
+    #   @param plugins [Array<Symbol>] A list of optional command plugins that should be used
+    #   @param meta [Hash] Meta data for a command
+    #
+    #   @return [Command, CommandProxy]
+    #
+    # @api private
+    def [](*args)
+      fetch_or_store(args.hash) do
```

Comment:
Oh I didn't know we freeze command classes. IIRC I *planned* to add freezing them (same with mappers and relations).